colossalai/shardformer/modeling/t5.py

```diff
@@ -582,6 +589,71 @@ def t5_encoder_model_forward(
 
         return outputs
 
+    @staticmethod
+    def t5_for_token_classification_forward(
+        self: T5ForTokenClassification,
+        input_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.FloatTensor] = None,
+        head_mask: Optional[torch.FloatTensor] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        stage_manager: Optional[PipelineStageManager] = None,
+        hidden_states: Optional[torch.FloatTensor] = None,
+        position_bias: Optional[torch.Tensor] = None,
+        encoder_decoder_position_bias: Optional[torch.Tensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        backward_tensor_keys: Optional[List[str]] = None,
+        stage_index: Optional[List[int]] = None,
+        decoder_starting_stage: Optional[int] = None,
+    ) -> Union[Tuple[torch.FloatTensor], BaseModelOutput]:
+        r"""
+        This function is modified on the basis of transformers.models.t5.modeling_t5.T5ForTokenClassification.forward.
+        Please refer to original code of transformers for more details.
+        ```"""
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        outputs = T5PipelineForwards.t5_stack_forward(
+            self.transformer.encoder,
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            inputs_embeds=inputs_embeds,
+            head_mask=head_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            stage_manager=stage_manager,
+            hidden_states=hidden_states,
+            position_bias=position_bias,
+            encoder_decoder_position_bias=encoder_decoder_position_bias,
+            stage_index=stage_index,
+            decoder_starting_stage=decoder_starting_stage,
+        )
+        if stage_manager.is_last_stage():
+            sequence_output = outputs[0]
+
+            sequence_output = self.dropout(sequence_output)
+            logits = self.classifier(sequence_output)
+
+            loss = None
+            if labels is not None:
+                loss_fct = CrossEntropyLoss()
+                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```

Comment:
If using tp, shall we use distributed cross entropy?

Reply:
I am not sure. This head is a token classification head instead of a casual LM head. For Existing Colossalai models, casual LM models in do split the cls head as well as cross entropy layer but Token Classification models (include Bert, GPT, T5, Bloom, Falcon) don't.  

If we use dist cross entropy layer here,  we need to add a parallel token classification head like **_VocabParallelLMHead1D_**, and a critical hyper-parameters “make_token_types_divisible_by” like _make_vocab_size_divisible_by_ than fix the existing token classification models(Bert, GPT, T5, Bloom, Falcon, ) . Should I do this? If so I will open a new PR to do this fix.

Reply:
I am not sure. This head is a token classification head instead of a casual LM head. For Existing Colossalai models, casual LM models in do split the cls head as well as cross entropy layer but Token Classification models (include Bert, GPT, T5, Bloom, Falcon) don't.  

If we use dist cross entropy layer here,  we need to add some methods and fix the existing token classification models(Bert, GPT, T5, Bloom, Falcon, ) . Should I do this? If so I will open a new PR to do this fix.

Reply:
The classifier is not sharded. No need to use dist cross entropy